raster/rt_core/rt_gdal.c

```diff
@@ -0,0 +1,217 @@
+ */
+int rt_raster_gdal_contour(
+	/* input parameters */
+	rt_raster src_raster,
+	int src_band,
+	int src_srid,
+	const char* src_srs,
+	double contour_interval,
+	double contour_base,
+	int fixed_level_count,
+	double *fixed_levels,
+	int polygonize,
+	/* output parameters */
+	size_t *ncontours,
+	struct rt_contour_t **contours
+	)
+{
+	CPLErr cplerr;
+	OGRErr ogrerr;
+	GDALRasterBandH hBand;
+	int nfeatures = 0, i = 0, j = 0;
+
+	_rti_contour_arg arg;
+	_rti_contour_arg_init(&arg);
+
+	/* Load raster into GDAL memory */
+	arg.src.ds = rt_raster_to_gdal_mem(src_raster, src_srs, NULL, NULL, 0, &(arg.src.drv), &(arg.src.destroy_drv));
+	/* Extract the desired band */
+	hBand = GDALGetRasterBand(arg.src.ds, src_band);
+
+	/* Set up the OGR destination data store */
+	arg.dst.srid = src_srid;
+	arg.dst.drv = OGRGetDriverByName("Memory");
+	if (!arg.dst.drv)
+		return _rti_contour_arg_destroy(&arg);
+
+	arg.dst.ds = OGR_Dr_CreateDataSource(arg.dst.drv, "contour_ds", NULL);
+	if (!arg.dst.ds)
+		return _rti_contour_arg_destroy(&arg);
+
+	/* Polygonize is 2.4+ only */
+#if POSTGIS_GDAL_VERSION >= 24
+	arg.dst.gtype = polygonize ? wkbPolygon : wkbLineString;
+#else
+	arg.dst.gtype = wkbLineString;
+#endif
+
+	/* Layer has geometry, elevation, id */
+	arg.dst.lyr = OGR_DS_CreateLayer(arg.dst.ds, "contours", NULL, arg.dst.gtype, NULL);
+	if (!arg.dst.lyr)
+		return _rti_contour_arg_destroy(&arg);
+
+	/* ID field */
+	OGRFieldDefnH hFldId = OGR_Fld_Create("id", OFTInteger);
+	ogrerr = OGR_L_CreateField(arg.dst.lyr, hFldId, TRUE);
+	if (ogrerr != OGRERR_NONE)
+		return _rti_contour_arg_destroy(&arg);
+
+	/* ELEVATION field */
+	OGRFieldDefnH hFldElevation = OGR_Fld_Create("elevation", OFTReal);
+	ogrerr = OGR_L_CreateField(arg.dst.lyr, hFldElevation, TRUE);
+	if (ogrerr != OGRERR_NONE)
+		return _rti_contour_arg_destroy(&arg);
+
+	// GDALContourGenerate( GDALRasterBandH hBand,
+	//   double dfContourInterval, double dfContourBase,
+	//   int nFixedLevelCount, double *padfFixedLevels,
+	//   int bUseNoData, double dfNoDataValue,
+	//   void *hLayer, int iIDField, int iElevField,
+	//   GDALProgressFunc pfnProgress, void *pProgressArg );
+
+	static int use_no_data = 0;
+	static double no_data_value = 0.0;
+
+	/* Run the contouring routine, filling up the OGR layer */
+	cplerr = GDALContourGenerate(
+		hBand,
+		contour_interval, contour_base,
+		fixed_level_count, fixed_levels,
+		use_no_data, no_data_value,
+		arg.dst.lyr, 0, 1, // OGRLayer, idFieldNum, elevFieldNum
+		NULL,              // GDALProgressFunc pfnProgress
+		NULL               // void *pProgressArg
+		);
+
+	if (cplerr != CE_None)
+		return _rti_contour_arg_destroy(&arg);
+
+	/* Convert the OGR layer into PostGIS geometries */
+	nfeatures = OGR_L_GetFeatureCount(arg.dst.lyr, TRUE);
+	if (nfeatures < 0)
+		return _rti_contour_arg_destroy(&arg);
+
+	*contours = rtalloc(sizeof(struct rt_contour_t) * nfeatures);
+	for (i = 0; i < nfeatures; i++) {
```

Comment:
iterating over features with OGR_L_GetFeature() is strongly discouraged as you make assumption on the values of the FID. Use rather OGR_L_GetNextFeature() and stop when NULL is returned